processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +56,313 @@
-          }
-          continue;
-        }
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
-        }
+      if (!queue.offer(new LoadNotice(lookupName, lookupExtractorFactoryContainer), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
       }
-      if (!failedExtractorFactoryMap.isEmpty()) {
-        throw new ISE(
-            "was not able to start the following lookup(s) [%s]",
-            failedExtractorFactoryMap.keySet().toString()
-        );
-      }
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup load notice", lookupName);
     }
   }
 
-  /**
-   * Add or update a lookup factory
-   *
-   * @param lookupName             The name of the lookup
-   * @param lookupExtractorFactory The factory of the lookup
-   *
-   * @return True if the lookup was updated, false otherwise
-   *
-   * @throws IllegalStateException if start of the factory fails
-   */
-  public boolean updateIfNew(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  public void remove(String lookupName)
   {
-    boolean update = false;
+    Preconditions.checkState(lifecycleLock.isStarted());
+
     try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
+      if (!queue.offer(new DropNotice(lookupName), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
+      }
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup drop notice", lookupName);
     }
-    try {
-      assertStarted();
-      LookupExtractorFactory prior = lookupMap.get(lookupName);
-      update = lookupExtractorFactory.replaces(prior);
-      if (update) {
-        if (!lookupExtractorFactory.start()) {
-          throw new ISE("Could not start [%s]", lookupName);
-        }
-        boolean racy;
-        do {
-          if (prior == null) {
-            racy = null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory);
-          } else {
-            racy = !lookupMap.replace(lookupName, prior, lookupExtractorFactory);
-          }
+  }
 
-          if (racy) {
-            prior = lookupMap.get(lookupName);
-            update = lookupExtractorFactory.replaces(prior);
-          }
-        } while (racy && update);
+  @Nullable
+  public LookupExtractorFactoryContainer get(String lookupName)
+  {
+    Preconditions.checkState(lifecycleLock.isStarted());
+    return lookupMap.get(lookupName);
+  }
 
-        if (prior != null && update) {
-          if (!prior.close()) {
-            LOGGER.error("Error closing [%s]:[%s]", lookupName, prior);
-          }
-        }
+  public LookupsState getAllLookupsState()
```

Comment:
Notice handling is idempotent if nothing happens in between and concurrently. If there is a duplicate drop notice, and it is applied after another load notice, the lookup will be removed unexpectedly.

I really don't want to try to understand what scenarios are possible and what are not, and how safe is it, so better just to make `getAllLookupsState()` consistent.

Since lookup add/remove traffic is supposed to be low, and generally the number of lookups is supposed to be small, I suggest the following solution: there is a class `LookupUpdateState`:
```java
class LookupUpdateState
{
  final ImmutableMap<String, LookupExtractorFactoryContainer> persisted;
  final ImmutableMap<String, Notice> pendingChanges;
  final ImmutableMap<String, Notice> changesBeingPersisted;
}
```

`LookupReferencesManager` has field `AtomicReference<LookupUpdateState> lookupUpdateStateRef`.

 - `get()` is `return lookupUpdateStateRef.get().persisted.get(lookupName)`
 - `add()` replaces and entry in `pendingChanges`, creates a new `LookupUpdateState`, replaces the reference of `lookupUpdateStateRef`. This is an atomic operation on `lookupUpdateStateRef`.
 - `remove()` - similar
 - mainThread
   1. Swaps `pendingChanges` to `changesBeingPersisted`, and `pendingChanges` with `ImmutableMap.of()` (empty).
   2. Applies the changes from `changesBeingPersisted` to `persisted`.
   3. Calls `takeSnapshot()`
   4. Replaces `persisted` with the updated map, and `changesBeingPersisted` with `ImmutableMap.of()`.

 - `getAllLookupsState()` - prepares the result in the required format from `lookupUpdateStateRef.get()`